From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Thomas Prescher <thomas.prescher@cyberus-technology.de>
Date: Fri, 26 Apr 2024 12:05:36 +0000
Subject: [PATCH] GIMHvOnKvm: don't pose as microsoft if virtio-gpu is in use

This is a workaround to have the Intel GPU driver load successfully which
does not work if we pose as microsoft.
---
 src/VBox/VMM/VMMR3/GIMHvOnKvm.cpp | 15 ++++++++++++---
 1 file changed, 12 insertions(+), 3 deletions(-)

diff --git a/src/VBox/VMM/VMMR3/GIMHvOnKvm.cpp b/src/VBox/VMM/VMMR3/GIMHvOnKvm.cpp
index b04c8441ac..362cc690f1 100644
--- a/src/VBox/VMM/VMMR3/GIMHvOnKvm.cpp
+++ b/src/VBox/VMM/VMMR3/GIMHvOnKvm.cpp
@@ -125,18 +125,27 @@ VMMR3_INT_DECL(int) gimR3HvInit(PVM pVM, PCFGMNODE pGimCfg)
             return rc2;
     }
 
+    /**
+     * If virtio-gpu is in use, revert back to VBoxVBoxVBox as HyperV Vendor because otherwise,
+     * the Intel GPU driver does not load.
+     */
+    bool withVirtioGPU {false};
+    int rc = CFGMR3QueryBoolDef(pCfgHv, "VirtioGPU", &withVirtioGPU, false);
+    AssertLogRelRCReturn(rc, rc);
+
     /** @cfgm{/GIM/HyperV/VendorID, string, 'VBoxVBoxVBox'}
      * The Hyper-V vendor signature, must be 12 characters. */
     char szVendor[13];
-    int rc = CFGMR3QueryStringDef(pCfgHv, "VendorID", szVendor, sizeof(szVendor), "Microsoft Hv");
+    rc = CFGMR3QueryStringDef(pCfgHv, "VendorID", szVendor, sizeof(szVendor), withVirtioGPU ? "VBoxVBoxVBox" : "Microsoft Hv");
     AssertLogRelRCReturn(rc, rc);
     AssertLogRelMsgReturn(strlen(szVendor) == 12,
                           ("The VendorID config value must be exactly 12 chars, '%s' isn't!\n", szVendor),
                           VERR_INVALID_PARAMETER);
 
-    LogRel(("GIM: HyperV: Reporting vendor as '%s'\n", szVendor));
+    AssertReleaseMsg(!RTStrNCmp(szVendor, GIM_HV_VENDOR_MICROSOFT, sizeof(GIM_HV_VENDOR_MICROSOFT) - 1) ||
+                     !RTStrNCmp(szVendor, GIM_HV_VENDOR_VBOX, sizeof(GIM_HV_VENDOR_VBOX) - 1), (("GIM Vendors other than Microsoft Hv and VBox are unsupported")));
 
-    AssertReleaseMsg(!RTStrNCmp(szVendor, GIM_HV_VENDOR_MICROSOFT, sizeof(GIM_HV_VENDOR_MICROSOFT) - 1), (("GIM Vendors other than Microsoft Hv are unsupported")));
+    LogRel(("GIM: HyperV: Reporting vendor as '%s'\n", szVendor));
 
     pHv->fIsInterfaceVs = false;
     pHv->fDbgHypercallInterface = false;
-- 
2.43.0

